### ISSUE TITLE
how to generate the negative samples for the scene_discriminator

### ISSUE THREAD

**Issue by leon532** [edenton/drnet-py#4]
`

    def train_scene_discriminator(x):
        netC.zero_grad()
        target = torch.cuda.FloatTensor(opt.batch_size, 1)
        x1 = x[0]
        x2 = x[1]
        h_p1 = netEP(x1).detach()
        h_p2 = netEP(x2).detach()

        half = int(opt.batch_size/2)
        rp = torch.randperm(half).cuda()
        h_p2[:half] = h_p2[rp]
        target[:half] = 1
        target[half:] = 0

        out = netC([h_p1, h_p2])
        bce = bce_criterion(out, Variable(target))

        bce.backward()
        optimizerC.step()

       acc =out[:half].gt(0.5).sum() + out[half:].le(0.5).sum()
       return bce.data.cpu().numpy(), acc.data.cpu().numpy()/opt.batch_size
`
I am confused how to generate the negative samples for the scene_discriminator. I think all items of the target vector should be 1.  

Any detailed explanation is welcome. Thank you~

**Comment by edenton** [comment]
The first half of the batch consists of pose vectors from the same video (and so the first half of the target vector is 1) and the second half of the batch consists of pose vectors from different videos (and so the target vector is 0). The pose vectors form different videos are constructed by just randomly permuting the order of the second pose vector (since each of the videos in a batch is likely to be different). Hope that clarifies!

**Comment by prashkmr** [comment]
I was going through this similar problem in identifying this and I am really sorry, but I am not able to get the explanation.

I changes the bach size to 8 and when I printed the x in
x = next(training_batch_generator)
output = [8,3,64,64]

When I print the x1 and x2 in the code
def train_scene_discriminator(x):
    netC.zero_grad()
    target = torch.cuda.FloatTensor(opt.batch_size, 1)
    x1 = x[0]
    x2 = x[1]
    print(x1)
    print(x2)    
I still get the same size 
[8,3,64,64]
[8,3,64,64]

I am literally confused about what the batch has and how is the scene_discriminator trained using frames from the same and different videos.
The training of the scene_discriminator is confusing me in how is the input (from same/different clip) assigned from x ?
Can you give some details of what x exactly has and how does all this workout ??


Thank you

**Comment by Feifannaner** [comment]
def train_scene_discriminator(x):
    # x shape(time_length, batch_size, channels, width, height) 
    x1=x[0] # the vector when time=0

**Comment by prashkmr** [comment]
@Feifannaner but the output is only 4 arguments.
@edenton commented above that 
"The first half of the batch consists of pose vectors from the same video (and so the first half of the target vector is 1) and the second half of the batch consists of pose vectors from different videos (and so the target vector is 0). The pose vectors form different videos are constructed by just randomly permuting the order of the second pose vector (since each of the videos in a batch is likely to be different)."

But how is the first half of the batch getting poses from the same and the second half from different?
Let's say in the above case when batch size is 8, we get 8 images but how is it that it has some images from same video and some from differnet(from which the different poses come).

Moreover although batch 'x' has 8 images, only 2 are used 
 x1 = x[0]
 x2 = x[1]

How can I even interpret the above comment by @edenton . with 
  x1 = x[0]
 x2 = x[1]


Please shed some light 

Thanks
Prashant

**Comment by Feifannaner** [comment]
Assuming  the batch_size=8,  and note that after the "randomperm" operation  in function "train_scene_discriminator", we could get h_p1 = 1,2,3,4,5,6,7,8, h_p2=1,2,3,4,(6,7,8,5);  then the first half the pose vectors are obviously from one scene(such as scene 1<-> scene 1) while the second half is not(scene 5 <-> scene 6).


For the 2nd question why only  2 images are used  in the function. I think it is solution to simplify the training process.

**Comment by prashkmr** [comment]
@Feifannaner I seem to get it now. Rechecked it and I see that the size of x is
[20,8,3,64,64], this seems to make some sense.

One clarification:
seq_length in the above is 20--> So if I am not wrong this means that the 1st index in the sequence has 8 (3*64*64) frames, where each frame is the first frame from 8 different videos

2nd index in the sequence has 8 (3*64*64) frames, where each frame is the second frame from 8 different videos

and this continues till 2- frames. 
Request you to give your opinion on the same.

**Comment by Feifannaner** [comment]
Sorry, I do not understand  the sentence "this continues till 2- frames." What is your question?

**Comment by Leong1230** [comment]
Hi! @Feifannaner
In train_drnet.py --> Function: train_scene_discriminator():

```
x1 = x[0]
x2 = x[1]
h_p1 = netEP(x1).detach()
h_p2 = netEP(x2).detach()

half = int(opt.batch_size/2)
rp = torch.randperm(half).cuda()
h_p2[:half] = h_p2[rp] # randomly permuting the first half of the vector
target[:half] = 1
target[half:] = 0
```

"h_p2[:half] = h_p2[rp]" seems to randomly permute the first half of the vector instead of the second half, so it should be "The first half of the batch consists of pose vectors from the **different** video and the second half of the batch consists of pose vectors from **same** videos ", and therefore:
```
target[:half] = 0
target[half:] = 1
```

Am I right or missing something in the code?

**Comment by Leong1230** [comment]
Also, for each training sample, 4 frames were randomly sampled from each video sequence, but I think maybe 2 frames from each sequence will be enough for calculating the losses, is that right? Since there were only 2 out of 4 frames were used in train_scene_discriminator(x).
If not, I wonder why the authors used different frames (i) to calculate Similarity Loss and Reconstruction Loss?

Your help would be greatly appreciated!
